res/controllers/common-controller-scripts.js

```diff
@@ -9,728 +9,618 @@
+     * Useful for manually synchronizing a track to an external beat.
+     * This function only works if the track's detected BPM value is correct.
+     * Call this function each time the tap button is pressed.
+     * @param {number} deck The deck to adjust.
+     */
+    tapButton(deck) {
+        const now = new Date().getTime() / 1000; // Current time in seconds
+        const tapDelta = now - this.tapTime;
+        this.tapTime = now;
+
+        // assign tapDelta in cases where the button has not been pressed previously
+        if (this.tap.length < 1) {
+            this.previousTapDelta = tapDelta;
         }
-        if (inputValue > currentRelative - this.maxJump
-            && inputValue < currentRelative + this.maxJump) {
-            engine.setValue(group, this.mappedFunction, outputValue);
+        // reset if longer than two seconds between taps
+        if (tapDelta > 2.0) {
+            this.tap = [];
+            return;
         }
-    } else {
-        engine.setValue(group, this.mappedFunction, outputValue);
-    }
-};
+        // reject occurrences of accidental double or missed taps
+        // a tap is considered missed when the delta of this press is 80% longer than the previous one
+        // and a tap is considered double when the delta is shorter than 40% of the previous one.
+        // these numbers are just guesses that produced good results in practice
+        if ((tapDelta > this.previousTapDelta * 1.8) || (tapDelta < this.previousTapDelta * 0.6)) {
+            return;
+        }
+        this.previousTapDelta = tapDelta;
+        this.tap.push(60 / tapDelta);
+        // Keep the last 8 samples for averaging
+        if (this.tap.length > 8) { this.tap.shift(); }
+        let sum = 0;
+        for (let i = 0; i < this.tap.length; i++) {
+            sum += this.tap[i];
+        }
+        const average = sum / this.tap.length;
 
-// Deck
+        const group = `[Channel${  deck  }]`;
 
-var Deck = function(deckNumber, group) {
-    this.deckNumber = deckNumber;
-    this.group = group;
-    this.Buttons = [];
-};
-Deck.prototype.setControlValue = Controller.prototype.setControlValue;
-Deck.prototype.addButton = Controller.prototype.addButton;
+        // "bpm" was changed in 1.10 to reflect the *adjusted* bpm, but I presume it
+        // was supposed to return the tracks bpm (which it did before the change).
+        // "file_bpm" is supposed to return the set BPM of the loaded track of the
+        // channel.
+        let fRateScale = average / engine.getValue(group, "file_bpm");
 
-// ----------------- END Object definitions ----------------------
+        // Adjust the rate:
+        fRateScale = (fRateScale - 1.) / engine.getValue(group, "rateRange");
+
+        engine.setValue(
+            group, "rate",
+            fRateScale * engine.getValue(group, "rate_dir")
+        );
+    }
+}
+// Add instance bpm of bpmClass to the Global JavaScript object
+this.bpm = new bpmClass();
```

Comment:
Why didn't you keep the IIFE in my fix?

Reply:
iiuc because it broke editor support for Joerg.

Reply:
What about using [JSDoc's `@global`](https://jsdoc.app/tags-global)?

Reply:
I just tried it again, with the proposed code the symbol bpm becomes invisible outside of the function statement. Neither code completion, pop-up help nor type checking work anymore. While with the code of this PR everything works.
`this.bpm = new bpmClass();` is not a new code pattern, it's already used in the hid-parser of 2.4 and seems to work for everybody.

Reply:
Weird. But ok then.